src/data/transform.ts

```diff
@@ -435,124 +427,159 @@ export const convertToExcalidrawElements = (
   if (!elements) {
     return [];
   }
+
   const elementStore = new ElementStore();
-  // Push all elements to array as there could be cases where element id
-  // is referenced before element is created
-  elements.forEach((element) => {
-    let elementId = element.id || regenerateId(null);
+
+  // ensure unique ids
+  // ---------------------------------------------------------------------------
+  const ids = new Set<ExcalidrawElement["id"]>();
+  const elementsWithIds = elements.map((element) => {
+    let id = element.id || regenerateId(null);
 
     // To make sure every element has a unique id since regenerateId appends
     // _copy to the original id and if it exists we need to generate again
     // hence a loop
-    while (elementStore.hasElementWithId(elementId)) {
-      elementId = regenerateId(elementId);
+    while (ids.has(id)) {
+      id = regenerateId(id);
     }
-    const elementWithId = { ...element, id: elementId };
-    elementStore.add(elementWithId as ExcalidrawElement);
+
+    ids.add(id);
+
+    return { ...element, id };
   });
+  // ---------------------------------------------------------------------------
 
-  const pushedElements =
-    elementStore.getElements() as readonly ExcalidrawProgrammaticElement[];
-  pushedElements.forEach((element) => {
-    const elementWithId = { ...element };
-
-    if (
-      (elementWithId.type === "rectangle" ||
-        elementWithId.type === "ellipse" ||
-        elementWithId.type === "diamond" ||
-        elementWithId.type === "arrow") &&
-      elementWithId?.label?.text
-    ) {
-      let [container, text] = bindTextToContainer(
-        elementWithId as
-          | ValidContainer
-          | ({
-              type: "arrow";
-            } & ValidLinearElement),
-        elementWithId?.label,
-      );
-      elementStore.add(container);
-      elementStore.add(text);
-
-      if (container.type === "arrow") {
-        const originalStart =
-          elementWithId.type === "arrow" ? elementWithId?.start : undefined;
-        const originalEnd =
-          elementWithId.type === "arrow" ? elementWithId?.end : undefined;
-        const { linearElement, startBoundElement, endBoundElement } =
-          bindLinearElementToElement(
-            {
-              ...container,
-              start: originalStart,
-              end: originalEnd,
-            },
-            elementStore,
+  for (const element of elementsWithIds) {
+    switch (element.type) {
```

Comment:
I've rewritten the looping into an exhaustive switch case. This way it's a little easier to read and we have more assurance we're not leaving things behind (especially in terms of implementation).

For this I've had to reorder some things around that were previously in specific if/else order. I don't think I've broken anything, but if I did then perhaps it's better it surfaced now instead of down the line. Lemme know.

Reply:
I wanted to have an array with allow bindable elements and check if type exists among it but it was failing in TS hence I added multiple "if" checks. Any better way to do in TS ?
eg
```
const bindingTypes = ["rectangle", "ellipse", "diamond", "arrow"];
if (bindingTypes.includes(elementWithId.type) {
    // Do stuff
}
```

Reply:
right now the most sure way to do exhaustive checks are switch cases.

Note that the aim is for TS to yell at us if we later introduce a new element type and forget to handle it here. This should now be happening.